tests/integrationv2/test_early_data.py

```diff
@@ -0,0 +1,397 @@
+S2N_EARLY_DATA_ACCEPTED_MARKER = S2N_EARLY_DATA_STATUS_MARKER.format(status="ACCEPTED")
+S2N_EARLY_DATA_REJECTED_MARKER = S2N_EARLY_DATA_STATUS_MARKER.format(status="REJECTED")
+
+
+class S2N(S2NBase):
+    def __init__(self, options: ProviderOptions):
+        S2NBase.__init__(self, options)
+
+    def setup_client(self):
+        cmd_line = S2NBase.setup_client(self)
+        early_data_file = self.options.early_data_file
+        if early_data_file and os.path.exists(early_data_file):
+            cmd_line.extend(['--early-data', early_data_file])
+        return cmd_line
+
+    def setup_server(self):
+        cmd_line = S2NBase.setup_server(self)
+        cmd_line.extend(['--max-early-data', self.options.max_early_data])
+        return cmd_line
+
+
+class OpenSSL(OpenSSLBase):
+    def __init__(self, options: ProviderOptions):
+        OpenSSLBase.__init__(self, options)
+
+    def setup_client(self):
+        cmd_line = OpenSSLBase.setup_client(self)
+        early_data_file = self.options.early_data_file
+        if early_data_file and os.path.exists(early_data_file):
+            cmd_line.extend(['-early_data', early_data_file])
+        ticket_file = self.options.ticket_file
+        if ticket_file:
+            if os.path.exists(ticket_file):
+                cmd_line.extend(['-sess_in', ticket_file])
+            else:
+                cmd_line.extend(['-sess_out', self.options.ticket_file])
+        return cmd_line
+
+    def setup_server(self):
+        cmd_line = OpenSSLBase.setup_server(self)
+        cmd_line.extend(['-early_data'])
+        return cmd_line
+
+
+# The `-reconnect` option is broken for TLS1.3 in OpenSSL s_client: https://github.com/openssl/openssl/issues/8517
+# The `-sess_in`/`-sess_out` options can be used instead, but don't have an s2nc equivalent.
+# As we add more providers, we may need both a `-reconnect`-like and a `-sess_in/out`-like S2N server test,
+# but for now we can just use `-sess_in/out` and cover the S2N->S2N case in the S2N client tests.
+CLIENT_PROVIDERS = [ OpenSSL ]
+SERVER_PROVIDERS = [ OpenSSL, S2N ]
+
+
+def get_early_data_bytes(file_path, early_data_size):
+    early_data = data_bytes(early_data_size)
+    with open(file_path, 'wb') as fout:
+        fout.write(early_data)
+    return early_data
+
+
+def get_ticket_from_s2n_server(options, managed_process, provider, certificate):
+    port = next(available_ports)
+
+    """
+    Generally clients start checking for stdin EoF to exit as soon as they finish the handshake.
+    To make sure the client reliably receives the post-handshake NST,
+    do NOT indicate stdin EoF until after some data has been received from the server.
+    """
+    close_marker_bytes = data_bytes(10)
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+    client_options.port = port
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+    server_options.port = port
```

Comment:
No, the port is actually being reassigned here. We use a separate port for the connection that establishes the ticket so that there's no contention possible when we then try to start up a separate server to actually use the ticket. 

(Not using the same server process lets us modify the server parameters like max_early_data_size after the ticket is issued)